### PR TITLE
support both endpoint modes in stack

### LINKED ISSUE #31096: Missing stack/compose support for `--endpoint-mode` option when creating a service.
**Description**

I want to use `docker stack deploy ...` to create some services which use the round-robin DNS endpoint mode. It doesn't look like it's possible. The only way to do it is to use `docker service create ...`.

Docker Version: 1.13.1

### PR DESCRIPTION
**Pull request by allencloud**
Signed-off-by: allencloud <allen.sun@daocloud.io>

fixes https://github.com/docker/docker/issues/31096

We found that stack currently has not yet supported dnsrr endpoint mode for service. This PR tries to fix that.

**- What I did**
1. add `EndpointMode` in `ServiceConfig` of compose/types
2. add `EndpointMode` in `Service` of bundlefile
3. make EndpointMode in stack passed to daemon side. 

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: cli/command/bundlefile/bundlefile.go [hunk 1]
```diff
@@ -14,15 +14,16 @@ type Bundlefile struct {
...
-	Ports      []Port            `json:",omitempty"`
-	WorkingDir *string           `json:",omitempty"`
-	User       *string           `json:",omitempty"`
-	Networks   []string          `json:",omitempty"`
+	Image        string
+	Command      []string          `json:",omitempty"`
+	Args         []string          `json:",omitempty"`
+	Env          []string          `json:",omitempty"`
+	Labels       map[string]string `json:",omitempty"`
+	EndpointMode *string           `json:",omitempty"`
```
Comment (dnephin) [inline cli/command/bundlefile/bundlefile.go]: Let's not add it to the bundlefile. The bundle is supposed to a subset of options, and I'm not sure it's appropriate.

**Comment by allencloud** [comment]
I think some test is needed as well, moving to WIP

File: cli/command/stack/deploy_bundlefile.go [hunk 2]
```diff
@@ -34,6 +36,11 @@ func deployBundle(ctx context.Context, dockerCli *command.DockerCli, opts deploy
 	for internalName, service := range bundle.Services {
 		name := namespace.Scope(internalName)
 
+		em := "vip"
+		if service.EndpointMode != nil {
+			em = *service.EndpointMode
+		}
```
Comment (dnephin) [inline cli/command/stack/deploy_bundlefile.go]: I don't think we should assume a default on the client. Different backends may want different defaults. Can we leave the default as `""` ?

File: cli/compose/convert/service.go [hunk 3]
```diff
@@ -374,7 +375,11 @@ func (a byPublishedPort) Len() int           { return len(a) }
 func (a byPublishedPort) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func (a byPublishedPort) Less(i, j int) bool { return a[i].PublishedPort < a[j].PublishedPort }
 
-func convertEndpointSpec(source []composetypes.ServicePortConfig) (*swarm.EndpointSpec, error) {
+func convertEndpointSpec(endpointMode string, source []composetypes.ServicePortConfig) (*swarm.EndpointSpec, error) {
+	if endpointMode == "" {
+		endpointMode = "vip"
+	}
```
Comment (dnephin) [inline cli/compose/convert/service.go]: Same comment here. We shouldn't assume a default in the client.

**Review by dnephin — COMMENTED** [review]
Thanks! Just a few comments

File: cli/command/bundlefile/bundlefile.go (hunk 1 above)
Comment (allencloud) [inline cli/command/bundlefile/bundlefile.go]: Thanks for your feedback. While, I was wondering if we do not specify `EndpointMode` in `bundlefile.Service`, there must be another way to specify endpoint mode for a service, isn't it?

Comment (dnephin) [inline cli/command/bundlefile/bundlefile.go]: If you're using bunldes you would have to `service update`, which is the case for many options.

**Comment by thaJeztah** [comment]
@dnephin looks like the PR was updated after your last review PTAL

**Comment by dnephin** [comment]
The change to bundlefile is still in. I think we need to exclude taht

**Comment by allencloud** [comment]
THANKS  for your feedback. @dnephin 
Removed that part, PTAL

**Review by shin- — COMMENTED** [review]
AFAICT, the 3.1 JSON schema needs to be updated as well.

**Comment by dnephin** [comment]
Oops, yes, you're right. The schema needs to be updated.

A test would be good too

**Comment by allencloud** [comment]
@dnephin @shin- 
Could you share a little bit more about `3.1 JSON schema`? Actually I have no idea about that.

**Comment by dnephin** [comment]
@allencloud the file is here: https://github.com/docker/docker/blob/master/cli/compose/schema/data/config_schema_v3.1.json

It's a http://json-schema.org/

**Comment by dnephin** [comment]
I pushed a commit that moves this field under `deploy`, and adds it to the jsonschema

**Comment by dnephin** [comment]
cc @vdemeester

**Comment by dnephin** [comment]
Carried in #31795

Thanks!

**Pull request closed without merging**